src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,686 @@
+
+        if(ret > 0)
+        {
+            /* Success */
+            plaintext_len += len;
+            return plaintext_len;
+        }
+        else
+        {
+			qCInfo(lcCse()) << "Error finalizing the decrypt";
+            /* Verify failed */
+            return -1;
+        }
+    }
+}
+
+ClientSideEncryption::ClientSideEncryption()
+{
+}
+
+void ClientSideEncryption::setAccount(AccountPtr account)
+{
+    _account = account;
+}
+
+void ClientSideEncryption::initialize()
+{
+    qCInfo(lcCse()) << "Initializing";
+    if (!_account->capabilities().clientSideEncryptionAvaliable()) {
+        qCInfo(lcCse()) << "No Client side encryption avaliable on server.";
+        emit initializationFinished();
+    }
+
+    if (hasPrivateKey() && hasPublicKey()) {
+        qCInfo(lcCse()) << "Public and private keys already downloaded";
+        emit initializationFinished();
+    }
+
+    getPublicKeyFromServer();
+}
+
+QString ClientSideEncryption::publicKeyPath() const
+{
+    return baseDirectory() + _account->displayName() + ".pub";
+}
+
+QString ClientSideEncryption::privateKeyPath() const
+{
+    return baseDirectory() + _account->displayName() + ".rsa";
+}
+
+bool ClientSideEncryption::hasPrivateKey() const
+{
+    return QFileInfo(privateKeyPath()).exists();
+}
+
+bool ClientSideEncryption::hasPublicKey() const
+{
+    return QFileInfo(publicKeyPath()).exists();
+}
+
+void ClientSideEncryption::generateKeyPair()
+{
+    // AES/GCM/NoPadding,
+    // metadataKeys with RSA/ECB/OAEPWithSHA-256AndMGF1Padding
+    qCInfo(lcCse()) << "No public key, generating a pair.";
+    const int rsaKeyLen = 2048;
+
+    EVP_PKEY *localKeyPair = nullptr;
+
+    // Init RSA
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+
+    if(EVP_PKEY_keygen_init(ctx) <= 0) {
+        qCInfo(lcCse()) << "Couldn't initialize the key generator";
+        return;
+    }
+
+    if(EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, rsaKeyLen) <= 0) {
+        qCInfo(lcCse()) << "Couldn't initialize the key generator bits";
+        return;
+    }
+
+    if(EVP_PKEY_keygen(ctx, &localKeyPair) <= 0) {
+        qCInfo(lcCse()) << "Could not generate the key";
+        return;
+    }
+    EVP_PKEY_CTX_free(ctx);
+    qCInfo(lcCse()) << "Key correctly generated";
+    qCInfo(lcCse()) << "Storing keys locally";
+
+    QDir dir;
+    if (!dir.mkpath(baseDirectory())) {
+        qCInfo(lcCse()) << "Could not create the folder for the keys.";
+        return;
+    }
+
+    auto privKeyPath = privateKeyPath().toLocal8Bit();
+    auto pubKeyPath = publicKeyPath().toLocal8Bit();
+    FILE *privKeyFile = fopen(privKeyPath.constData(), "w");
+    FILE *pubKeyFile = fopen(pubKeyPath.constData(), "w");
+
+    qCInfo(lcCse()) << "Private key filename" << privKeyPath;
+    qCInfo(lcCse()) << "Public key filename" << pubKeyPath;
+
+    //TODO: Verify if the key needs to be stored with a Cipher and Pass.
+    if (!PEM_write_PrivateKey(privKeyFile, localKeyPair, NULL, NULL, 0, 0, NULL)) {
```

Comment:
We should probably not write it as PEM but rather as DER (base64 encoded) when sending it to the server so it is all on 1 line.